app/models/user.rb

```diff
@@ -48,4 +49,22 @@ class User < ApplicationRecord
 
   scope :ordered, -> { order(username: :asc) }
 
+  def self.new_with_session(params, session)
+    super.tap do |user|
+      if data = session["devise.facebook_data"] && session["devise.facebook_data"]["extra"]["raw_info"]
+        user.email = data["email"] if user.email.blank?
+      end
+    end
+  end
+
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.email = auth.info.email
+      user.password = Devise.friendly_token[0,20]
+      user.username = auth.info.name
+      user.remote_avatar_url = auth.info.image
```

Comment:
for facebook please use `JSON.load(open("#{auth[:info][:image]}?redirect=false&type=large"))['data']['url']`